### PR TITLE
feat: profanity filter & added "shuttle.rs" to the reserved list of project names

### PR DESCRIPTION
**Pull request by Butch78**
This ensures users aren't using profanity for the names of their shuttle projects. It also creates an easy way to add to the reserved project name list.

### CODE REVIEW

File: common/src/project.rs [hunk 1]
```diff
@@ -124,6 +144,8 @@ pub mod tests {
             ".invalid",
             "invalid.name",
             "invalid.name.",
+            "test-fuck-fuck",
```
Comment (christoshadjiaslanis) [inline common/src/project.rs:147]: consider changing to `crap` as it is still profanity and little less abrasive :smiley:

File: common/src/project.rs [hunk 2]
```diff
@@ -41,14 +47,26 @@ impl std::fmt::Display for ProjectName {
...
 
-        !(hostname.bytes().any(|byte| !is_valid_char(byte))
+        fn is_profanity_free_and_not_reserved(hostname: &str) -> bool {
+
+            let mut reserved_words = HashSet::new();
+            // Reserved words can be easily added here. 
+            reserved_words.insert("shuttle.rs".to_string());
+
+
+            let censor = Censor::Standard + Censor::Sex + Censor::Custom(reserved_words);
```
Comment (chesedo) [inline common/src/project.rs:65]: Maybe this can be replaced with an `once_cell` to not initialize it each time?

File: common/src/project.rs [hunk 3]
```diff
@@ -41,14 +47,26 @@ impl std::fmt::Display for ProjectName {
...
                 || (b'0'..=b'9').contains(&byte)
                 || byte == b'-'
         }
 
-        !(hostname.bytes().any(|byte| !is_valid_char(byte))
+        fn is_profanity_free_and_not_reserved(hostname: &str) -> bool {
+
+            let mut reserved_words = HashSet::new();
+            // Reserved words can be easily added here. 
+            reserved_words.insert("shuttle.rs".to_string());
```
Comment (chesedo) [inline common/src/project.rs:62]: Creating the `HashSet` from an array ([the last doc example](https://doc.rust-lang.org/std/collections/struct.HashSet.html)) may make it even easier to add reserve words in the future :man_shrugging:

**Review by chesedo — COMMENTED** [review]
Wow, nice and simple! Thanks @Butch78

File: common/src/project.rs (hunk 2 above)
Comment (Butch78) [reply common/src/project.rs]: Fixed! :)

**Review by chesedo — APPROVED** [review]
Nice one @Butch78!

**Comment by chesedo** [comment]
@Butch78 it seems we only need to run a `cargo fmt --all` here

File: common/src/project.rs [hunk 4]
```diff
@@ -48,7 +56,19 @@ impl ProjectName {
...
+
+            let censor = Censor::Standard
+                + Censor::Sex
+                + Censor::Custom(INSTANCE.get().expect("Reserved words not set").clone());
+            return !censor.check(hostname);
+        }
+
+        !(hostname
+            .bytes()
+            .any(|byte| !is_valid_char(byte) | !is_profanity_free_and_not_reserved(hostname))
```
Comment (coszio) [inline common/src/project.rs:71]: I think you meant to add the check outside of the `any` closure to only check it once?

File: common/src/project.rs [hunk 5]
```diff
@@ -48,7 +56,19 @@ impl ProjectName {
...
             || hostname.ends_with('-')
```
Comment (chesedo) [inline common/src/project.rs:72]:
```suggestion
            .any(|byte| !is_valid_char(byte))
            || !is_profanity_free_and_not_reserved(hostname)
            || hostname.ends_with('-')
```
I think this is what @coszio means. That way the same `is_profanity_free_and_not_reserved(hostname)` is not made with every byte check

File: common/src/project.rs (hunk 4 above)
Comment (Butch78) [reply common/src/project.rs:71]: Agreed good save @coszio 👍

File: common/src/project.rs (hunk 5 above)
Comment (Butch78) [reply common/src/project.rs:72]: Easy fix, I'll do it now! :)

**Comment by chesedo** [comment]
@Butch78 we just need to merge in the underscore check that was merged to `main` recently

**Pull request merged**